### ISSUE TITLE
Single-seq decode CUDA graph destroyed on free_sequence → unnecessary re-capture per slot occupant (all models)

### ISSUE THREAD

**Issue by rsafier** [Avarok-Cybersecurity/atlas#323]
## Summary

The single-sequence decode **CUDA graph is destroyed on every sequence free** (`free_sequence`), even though the graph is fully **reusable across any occupant of the same slot**. As a result each new sequence that reuses a slot re-captures an identical graph on its first decode step. This is generic model infrastructure (`decode_graph`), so **it affects every model** that uses the single-seq (C=1 / batch-drains-to-1) decode path.

Observed empirically on Qwen3.6-27B-NVFP4 (GB10): over a ~33-min mixed C=1..8 run, `slot=0` was captured **8×** and `slot=1` **2×** (10 captures total) — i.e. slots repeatedly go cold and re-capture, rather than capturing once and persisting.

## Mechanism

The single-seq decode graph cache `decode_graph: HashMap<slot_idx, GraphHandle>` is:

- **Replayed** slot-keyed (`crates/spark-model/src/model/trait_impl/decode_a.rs`, Phase 2): `cache.get(&seq.slot_idx)` → `launch_graph`. All per-step dynamic data is read from device staging refreshed *before* replay.
- **Captured** on a cold-slot miss (decode_a.rs Phase 3): `end_capture` → `cache.insert(seq.slot_idx, graph)`.
- **Destroyed on free** (`crates/spark-model/src/model/trait_impl/sequence.rs`, `free_sequence`):
  ```rust
  // Invalidate cached CUDA graphs that reference this sequence's slot
  // — the graph was captured with this slot's KV/SSM pointers baked in,
  // and replaying after the slot is freed would read stale data.
  if let Some(graph) = self.decode_graph.lock().remove(&seq.slot_idx) { destroy_graph(graph) }
  ```

So under normal request turnover — and *maximally* under variable-length co-dispatch, where sequences finish at staggered times and free slots constantly — every slot reuse pays a fresh capture (a full 64-layer build + non-overlappable sync).

## Why the "stale data" rationale does not hold

Nothing baked into the graph is unique to a specific sequence beyond its **slot**:

| What varies per step / per sequence | How it reaches the kernel | Baked into graph? |
|---|---|---|
| positions, slot, seq_len, block_table, token_id | uploaded every step to a **fixed** `meta_base = buffers.scratch().offset(32768)` staging region (decode_a.rs Phase 1) | No — refreshed pre-replay |
| SSM `h_state` / `conv_state` | `ssm_pool.h_state(layer, slot)` — a pure function of **(layer, slot)** | Pointer baked, but **identical for any occupant of that slot** |
| hidden / residual / KV base / weights | fixed model buffers | Same for all sequences |

Two independent proofs the graph is slot-reusable, not sequence-specific:

1. **Empirical, already true today:** one capture serves all ~256 decode steps of a sequence while its `seq_len`, positions, and `block_table` all change every step. That proves everything varying *within* a sequence's life is device-staging-refreshed, not baked.
2. **Static:** the only per-occupant pointer is the SSM state address, and it is slot-relative (`h_state(layer, slot)`), so it is the same for any sequence in that slot. `compact_sequence` repoints migrated survivors to `h_state(layer, new_slot)` consistently.

Replay only ever occurs in `decode()` for a **live** occupant of the slot, whose state (written to the slot's fixed pool address by its prefill) and metadata (uploaded to fixed staging pre-replay) are both fresh. The `free_sequence` zeroing of `seq.layer_states[i].h_state = DevicePtr(0)` only nulls the `SequenceState` struct's pointer copies — it does not touch the pool arena the graph baked. So the "would read stale data" concern does not correspond to any reachable control flow.

## Proposed fix

In `free_sequence`, **do not remove/destroy `decode_graph[seq.slot_idx]`** — let it persist for the slot's next occupant; destroy graphs only at model teardown. Memory is bounded (≤ `--max-num-seqs` graphs, one per slot).

**Scope guardrails:**
- Applies to `decode_graph` (single-seq, slot-keyed) and, by the same argument, the `verify2/3/4_graph` slot-keyed caches.
- **Does NOT apply** to `batch_decode_graphs` (keyed by `padded_n`), which bakes the *whole active set's* per-slot pointers and is correctly drained on free.
- One item needs eyes-on before shipping: whether the **cuBLASLt-in-graph** path (`ATLAS_CUBLAS_GEMM=1`) bakes a per-call workspace/algo pointer. If workspace is a fixed model buffer (typical), fine; if per-sequence, that slot's graph would still need to be dropped.

## Validation required (needs GPU)

- **Bit-identical logits:** capture-once → replay for a *fresh* slot occupant must produce identical logits to the current re-capture path.
- **Needle / long-generation** run to confirm no cross-occupant contamination.
- Confirm capture count **plateaus** at ≤ #slots (vs. the current growth of 8×/slot).

## Impact / priority

Per-capture cost is ~1.5–5 ms (64-layer build + sync). Negligible for long generations; real only under **high turnover of short sequences at low concurrency**. This is a **design-cleanliness / wasted-work** fix, not a major throughput lever — but it is cheap, low-risk, removes non-overlappable sync spikes, and benefits all models.

## References (paths at the current tree)

- `crates/spark-model/src/model/trait_impl/decode_a.rs` — Phase 1 fixed-staging upload; Phase 2 slot-keyed replay; Phase 3 capture+insert; comment "the captured graph works for any max_blocks … seq_idx=0 makes the stride irrelevant".
- `crates/spark-model/src/model/trait_impl/sequence.rs` — `free_sequence` remove+destroy of `decode_graph[slot]`; `compact_sequence_dispatch` repoint to `h_state(layer, new_slot)`.
- `ssm_pool.h_state(layer, slot)` / `conv_state(layer, slot)` — slot-relative addressing.

---
_Discovered during concurrency benchmarking of Qwen3.6-27B-NVFP4 on branch `worktree-q12-flag-eager-n-fix` (base commit `287e73d3`). The affected code (`decode_a.rs`, `sequence.rs`) is **unmodified** by that branch — this is a pre-existing, main-line behavior independent of the Q12/eager-N work on that branch._
